[Feature] Add Simulated Annealing (#60)

NavidCOMSC opened the pull request:
The developments break down into two classes, written in two different files that with initializer makes up the metahueristic module.
The ``JobShopAnnealer`` class demonstrates the developments of the simulated annealing methodology. It consists of method to represent the current state of the model after each search and evaluate its acceptance criteria. ``move`` method makes a random swap of operations and ``energy`` method computes the current state of the solution in the comparison with the solution from the former state. It imposes the penalty and computes its value by the constraints of ``deadlines`` and ``arrival_times``. The class tries to optimize for the solution makespan and find the next available operation.
The ``SimulatedAnnealingSolver`` class inherits from the abstract class of ``BaseSolver``. It computes the the best feasible solution after generating the initial state.
Tests are written a single file covering both classes.

Pabloo22 reviewed (changes requested):
Hi, I hope this review also answers your question in the issue comment.

The main confusion was that you forgot I added `release_dates`, `deadlines`, and `due_times` attributes to the `Operation` class in a recent version. I left a comment on how to actually use these attributes when creating an instance. The `JobShopInstance` class reads these operation attributes and infers their respective matrices. You also have properties to check whether an instance has release dates, deadlines, or due times.

NavidCOMSC commented on job_shop_lib/dispatching/rules/_dispatching_rule_solver.py:
```diff
@@ -56,7 +56,7 @@ class DispatchingRuleSolver(BaseSolver):
               using the specified filters.
 
-    .. seealso::
+    .. see also::
```
my raw inexperience with Sphinx documentation! Remove the space.

NavidCOMSC commented on job_shop_lib/metaheuristics/__init__.py:
```diff
@@ -0,0 +1,19 @@
+"""
+Contains the implementation of the JobShopAnnealer class,
+which is a simulated annealing algorithm for solving job
+shop scheduling problems.
+
+SA (Simulated Annealing) is a probabilistic technique (metaheuristic
+algorithm) for approximating the global optimum of a given function.
+It seeks to find a good approximation of the optimized makespan
+by exploring the schedule space and accepting worse schedules with
+evaluating the metropolis criterion. It represents the JSSP solution by
+defining the objective function that complies with the constraints of the
+arrival times and deadlines.
+"""
```
Yeah, I overlooked the fact the initializer can be taken as a more generic place for other metaheuristic algorithms. I have replaced the docstring that shows the generalizability of the metaheuristic module.

NavidCOMSC commented on job_shop_lib/metaheuristics/__init__.py:
```diff
@@ -0,0 +1,19 @@
+arrival times and deadlines.
+"""
+
+from job_shop_lib.metaheuristics._job_shop_annealer import JobShopAnnealer
```
I imported both classes in here

Pabloo22 commented on pyproject.toml:
```diff
@@ -21,6 +21,7 @@ imageio = {extras = ["ffmpeg"], version = "^2.34.1"}
 pygraphviz = {version = "^1.12", optional = true}
 numpy = ">=1.26.4,<3.0.0"
 gymnasium = "^1.0.0"
+simanneal = { git = "https://github.com/perrygeo/simanneal.git", rev = "master" }
```
Hi @NavidCOMSC,

Yes, use `poetry add` please. The problem with fetching the latest merges to the main branch is that it could cause compatibility problems in the future. Installing from pypi guarantees that we never install a new major version with backward compatibility issues inadvertently.

NavidCOMSC replied:
All done!

NavidCOMSC commented on job_shop_lib/metaheuristics/_job_shop_annealer.py:
```diff
@@ -0,0 +1,139 @@
+
+from job_shop_lib import JobShopInstance, Schedule
+from job_shop_lib import Schedule as ScheduleBuilder
+from job_shop_lib import ScheduledOperation
```
merged all import in one line

NavidCOMSC commented on job_shop_lib/metaheuristics/_job_shop_annealer.py:
```diff
@@ -0,0 +1,139 @@
+        """Calculates penalties for the constraints of
+        deadline and arrival time violations."""
+        if not hasattr(self.instance, "deadlines") and not hasattr(
+            self.instance, "arrival_times"
```
@Pabloo22  I have edited to compute penalty function to use the attribute of ``release_date_matrix`` and property checker of ``has_release_dates`` instead. I am uncertain whether we should use the property of ``has_due_times`` at all? Please comment to clarify. Thanks.

NavidCOMSC replied:
I rather not add an extra changes as I have been through a hectic debugging that I could fix all test. By adding this extra penalty for the property of ``due_times`` by publishing a separate issue to handle its condition. We can speak over it when it comes to terms of adding it to the annealing class.

Pabloo22 commented on tests/metaheuristics/test_sa_solver.py:
```diff
@@ -0,0 +1,154 @@
+from job_shop_lib import Schedule
+
+
+# Fixture for a simple job shop instance
```
@NavidCOMSC this comment will help you the most probably

NavidCOMSC replied:
changes applied.

NavidCOMSC commented on tests/metaheuristics/test_sa_solver.py:
```diff
@@ -0,0 +1,154 @@
+
+
+# Fixture for ft06 instance
+@pytest.fixture
```
Commented out of the test file!

NavidCOMSC commented on tests/metaheuristics/test_sa_solver.py:
```diff
@@ -0,0 +1,154 @@
+        assert completion_time <= simple_instance.deadlines[job_id]
+
+
+# Import test
```
Import test is removed

Pabloo22 commented on job_shop_lib/metaheuristics/_job_shop_annealer.py:
```diff
@@ -0,0 +1,146 @@
+            return makespan + penalty
+        except ValidationError as e:
+            # If the schedule is invalid, return a large penalty
+            logging.warning("Invalid schedule encountered: %s", e)
```
I'd remove this warning to avoid spamming users.

Pabloo22 commented:
Hi, regarding your question in issue #60, it appears the test fails because the solver was unable to find a feasible schedule after completing the specified number of steps. A possible solution could be to increase the number of steps (if the running time is not too large), use an initial state obtained using a dispatching rule, or both.

I suggest starting with an initial state obtained from the "most work remaining" rule, and then increasing the number of steps if needed. The makespan after applying the rule is 61, so checking if it goes below 61 would be perfect. Also, could you modify the class so that it accepts a seed as an argument? This would be great to ensure the same results are reproduced every time we run the tests. In research, this is typically very important too.

By the way, I'd remove the warning raised when the schedule is moved and results in an unfeasible solution (I added a comment in the relevant line).

NavidCOMSC commented on job_shop_lib/metaheuristics/_simulated_annealing_solver.py:
```diff
@@ -0,0 +1,79 @@
+        self,
+        instance: JobShopInstance,
+        initial_state: list[list[int]] | None = None,
+    ) -> Schedule:
```
completed

NavidCOMSC commented on job_shop_lib/metaheuristics/_simulated_annealing_solver.py:
```diff
@@ -0,0 +1,79 @@
+
+    def __init__(
+        self,
+        initial_temperature: float = 10000.0,
```
the dictionary parameters are added to the class constructor that enhance users with more flexibility in selecting and initializing parameters for the solution of simanneal library.

Pabloo22 commented on job_shop_lib/metaheuristics/__init__.py:
Change this to use ".. autosummary"

NavidCOMSC replied:
I have upgraded the docstring of the class. Does it look how you were saying to be?

Pabloo22 commented on job_shop_lib/metaheuristics/__init__.py:
```diff
@@ -0,0 +1,34 @@
+algorithms are nature-inspired or probabilistic techniques that can find
+near-optimal solutions for complex combinatorial optimization problems.
+
+Metaheuristics are particularly well-suited for JSSP due to their ability to:
```
We need an extra space after this so that it renders the bullet points correctly

Pabloo22 commented on job_shop_lib/metaheuristics/__init__.py:
We need an extra space after this so that it renders the bullet points correctly

Pabloo22 commented on job_shop_lib/metaheuristics/_job_shop_annealer.py:
```diff
@@ -0,0 +1,144 @@
+
+        if self.instance.has_deadlines:
+            # Use the deadlines_matrix property to get deadlines
+            for job_id, completion_time in enumerate(job_completion_times):
+                deadlines = self.instance.deadlines_matrix[job_id]
+                # Check the last operation's deadline for each job
+                last_op_deadline = deadlines[-1]
+                if (
+                    last_op_deadline is not None
+                    and completion_time is not None
+                    and completion_time > last_op_deadline
+                ):
+                    penalty += float(self.penalty_factor) * (
+                        completion_time - last_op_deadline
+                    )
+
+        if self.instance.has_release_dates:
+            arrival_times = self.instance.release_dates_matrix
+            for job_id in range(self.instance.num_jobs):
+                # Check for the first operation's release date for each job
+                first_op_release_date = arrival_times[job_id][0]
+                if first_op_release_date is not None:
+                    scheduled_op = self._find_scheduled_operation(
+                        schedule, job_id, 0
+                    )
+                    if scheduled_op.start_time < first_op_release_date:
+                        penalty += float(self.penalty_factor) * (
+                            first_op_release_date - scheduled_op.start_time
+                        )
```
Penalties should check all operations, not only the last or first ones. Sorry, maybe I oversaw this in my previous review.

Pabloo22 commented on job_shop_lib/metaheuristics/_job_shop_annealer.py:
```diff
@@ -0,0 +1,144 @@
+        """Returns the completion time of the last operation for each job."""
+        completion_times: list[int | None] = [None] * self.instance.num_jobs
+
+        for machine_schedule in schedule.schedule:
+            for operation in machine_schedule:
+                job_id = operation.job_id
+                # Check if this the last operation for this job
+                if (
+                    operation.position_in_job
+                    == self.instance.num_operations - 1
+                ):
+                    completion_times[job_id] = operation.end_time
+
+        return completion_times
```
Here, it's sufficient and significatnly more efficient to only access the last operation of each job

Pabloo22 commented on job_shop_lib/metaheuristics/_simulated_annealing_solver.py:
```diff
@@ -0,0 +1,158 @@
+        # Save current random state and set new seed if provided
+        if self.seed is not None:
+            old_state = random.getstate()
+            random.seed(self.seed)
```
I think I did the same error of trying to set the seed this way in the `GeneralInstanceGenerator` class. I need to check it, but this probably doesn't work. We need to use a `np.RandomState` or `random.Random` classes.

NavidCOMSC replied:
@Pabloo22 I have made some modifications to the random generator that uses a local generator and if the seed is not given falls back to the global generator, so it is backward compatible.
I am closing this comment, though it is at your disposal if you are planning to upgrade the whole random generator approach.

NavidCOMSC commented:
Hi @Pabloo22, I have addressed some of the reviews you have left after raising the PR. I tried to leave you with the adequate description for reasoning behind the approach of each change. If you already noticed, I leave out two of the reviews for you to tackle, because the current format of the implementation is conducive to tests and their conditional checks. The last push adheres with the all tests and linting checks. Please give me a nudge if you need any help. Cheers.

Pabloo22 commented:
@NavidCOMSC Hi, did you change  `simanneal`'s default initial temperature from 25000.0 to 10000.0 for some reason? Same thing with steps (50000 --> 10000).

Pabloo22 commented:
/gemini review
